File: src/main/java/com/github/tuk-cps/ExprParser/ExprTreeFunction.java
package ExprParser;

import jAADD.AADD;
import jAADD.BDD;
import jAADD.DD;

import java.util.ArrayList;
import java.util.Arrays;

/**
 * A function call of a user-defined function.
 * It defines a local scope by mapping the parameters to the variables used during declaration of the function.
 * It searches for variables only in the local scope.
 */
class ExprTreeFunction extends ExprTree {
    protected ArrayList<ExprTree> param;
    protected  ExprTree   ast;

    ExprTree getAst()             { return ast; }
    void     setAst(ExprTree ast) { if (ast!=null) this.ast = ast.copy(); }

    ExprTreeFunction()  { super(); param = new ArrayList<>(); }
    ExprTreeFunction(DD val, String name, ExprTree ... param)  {
        super(name, val);
        this.ast= null;
        this.param = new ArrayList<ExprTree>(Arrays.asList(param));
    }

    @Override
    ExprTree copy() {
        try {
            ExprTreeFunction cp = getClass().getDeclaredConstructor().newInstance();
            cp.setId(getId());
            cp.setResult(getResult());
            cp.setAst(getAst());
            cp.param = new ArrayList<>(param);
            for (int i = 0; i < param.size(); i++)
                cp.param.set(i, param.get(i).copy());
            return cp;
        } catch(Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    /**
     * Processes a user-defined function call.
     * First, gets the function from the symbol table.
     * @throws ExprError
     */
    @Override
    void evalThis() throws ExprError {
        assert id != null;
        assert ast != null : "AST of function call must not be null";
        // throw new ExprError("function "+id+" not defined."); // function not defined
        // ast.evalExpr();
        // ast = getSymTab().getFuncCall(id, param);
        ast.evalExpr();
        result = ast.result;
    }


    // Computes the complete tree recursively.
    @Override
    void evalExpr() throws ExprError {
        assert param != null: "expected parameters are null";
        getSymTab().scopes.push(this);

        for (ExprTree par: param) { par.evalExpr(); }
        evalThis();
        getSymTab().scopes.pop();
    }


    ExprTree getParam(int nr) throws ExprError {
        if (nr >= param.size()) throw new ExprError("not enough parameters");
        return param.get(nr);
    }


    BDD getBDDParam(int nr) throws ExprError {
        if (nr >= param.size()) throw new ExprError("not enough parameters");
        return param.get(nr).getBDD();
    }


    AADD getAADDParam(int nr) throws ExprError {
        if (nr >= param.size()) throw new ExprError("not enough parameters");
        return param.get(nr).getAADD();
    }


    @Override
    public String toString() {
        try {
            String s = "( "+ getResult() + " "+ id;
            if (param != null)
                for (ExprTree par : param) {
                    s += par + " ";
                }
            if (ast != null)
                s += " -> " + ast;
            else
                s += " -> builtin ";
            s+=")";
            return s;
        } catch  (Exception e) {
            return "Error printing function call";
        }
    }
}

File: src/main/kotlin/com/github/tuk-cps/exprParser/ExprTreeFunction.kt
package exprParser

import jAADD.AADD
import jAADD.BDD
import jAADD.DD
import java.util.*

/**
 * A function call of a user-defined function.
 * It defines a local scope by mapping the parameters to the variables used during declaration of the function.
 * It searches for variables only in the local scope.
 */
open class ExprTreeFunction : ExprTree {

    var param: ArrayList<ExprTree> = ArrayList<ExprTree>()
    internal var ast: ExprTree? = null
        set(value) { field = if (value==null) null else value.copy() }

    internal constructor()

    /**
     * @method constructor
     * @param val
     * @param name
     * @param param
     */
    internal constructor(value: DD<*>, name: String, vararg param: ExprTree) : super(value, name) {
        this.param = ArrayList<ExprTree>(listOf(*param))
        // ToDo: Iterate over parameters, set parent pointer.
        // cannot set parent pointers to ExprTree as it has no children
        for (currtree in this.param) {
            currtree.parent = null
            //addPP(currtree);
        }
    }


    public override fun copy(): ExprTree {
        return try {
            val cp = javaClass.getDeclaredConstructor().newInstance()
            cp.parent = parent
            cp.id = id
            cp.value = value
            cp.ast = ast
            cp.param = ArrayList(param)
            for (i in param.indices) cp.param[i] = param[i].copy()
            cp
        } catch (e: Exception) {
            e.printStackTrace()
            return ExprTree(BDD.False, "Exception")
        }
    }

    /**
     * Processes a user-defined function call.
     * First, gets the function from the symbol table.
     * @throws ExprError
     */
    @Throws(ExprError::class)
    public override fun evalUp() {
        assert(ast != null) { "AST of function call must not be null" }
        // throw new ExprError("function "+id+" not defined."); // function not defined
        // ast.evalExpr();
        // ast = getSymTab().getFuncCall(id, param);
        ast!!.evalUpRec()
        value = ast!!.value
    }

    // Computes the complete tree recursively.
    @Throws(ExprError::class)
    public override fun evalUpRec() {
        symTab.scopes.push(this)
        for (par in param) {
            par.evalUpRec()
        }
        evalUp()
        symTab.scopes.pop()
    }

    @Throws(ExprError::class)
    fun getParam(nr: Int): ExprTree {
        if (nr >= param.size) throw ExprError("not enough parameters")
        return param[nr]
    }

    @Throws(ExprError::class)
    fun getBDDParam(nr: Int): BDD {
        if (nr >= param.size) throw ExprError("not enough parameters")
        return param[nr].bdd
    }

    @Throws(ExprError::class)
    fun getAADDParam(nr: Int): AADD {
        if (nr >= param.size) throw ExprError("not enough parameters")
        return param[nr].aadd
    }

    override fun toString(): String {
        return try {
            var s = "( " + value + " " + id
            for (par in param) {
                s += "$par "
            }
            s += if (ast != null) " -> $ast" else " -> builtin "
            s += ")"
            s
        } catch (e: Exception) {
            "Error printing function call"
        }
    }
}